File: frontend/lib/models/dedup.dart
import 'package:equatable/equatable.dart';

enum DedupWarningStatus {
  pending, dismissed, merged;

  static DedupWarningStatus fromString(String s) => switch (s) {
    'pending' => pending,
    'dismissed' => dismissed,
    'merged' => merged,
    _ => pending,
  };
}

final class DedupWarning extends Equatable {
  const DedupWarning({
    required this.id,
    required this.resourceIdA,
    required this.resourceIdB,
    required this.similarityScore,
    required this.status,
  });

  factory DedupWarning.fromJson(Map<String, dynamic> json) => DedupWarning(
    id: json['id'] as String,
    resourceIdA: json['resource_id_a'] as String,
    resourceIdB: json['resource_id_b'] as String,
    similarityScore: (json['similarity_score'] as num).toDouble(),
    status: DedupWarningStatus.fromString(json['status'] as String),
  );

  final String id;
  final String resourceIdA;
  final String resourceIdB;
  final double similarityScore;
  final DedupWarningStatus status;

  @override
  List<Object?> get props => [id, resourceIdA, resourceIdB, similarityScore, status];
}

final class MergeInput {
  const MergeInput({required this.keepId, required this.discardId});

  final String keepId;
  final String discardId;

  Map<String, dynamic> toJson() => {
    'keep_id': keepId,
    'discard_id': discardId,
  };
}

File: backend/domain/src/dedup.rs
use async_trait::async_trait;
use chrono::{DateTime, Utc};
use uuid::Uuid;

use crate::DomainError;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum DedupWarningStatus {
    Pending,
    Dismissed,
    Merged,
}

#[derive(Debug, Clone)]
pub struct DedupWarning {
    pub id: Uuid,
    pub resource_id_a: Uuid,
    pub resource_id_b: Uuid,
    pub similarity_score: f64,
    pub status: DedupWarningStatus,
    pub created_at: DateTime<Utc>,
    pub resolved_at: Option<DateTime<Utc>>,
}

pub struct NewDedupWarning {
    pub resource_id_a: Uuid,
    pub resource_id_b: Uuid,
    pub similarity_score: f64,
}

#[async_trait]
pub trait DedupWarningRepository: Send + Sync {
    async fn create(&self, warning: NewDedupWarning) -> Result<DedupWarning, DomainError>;
    async fn list_pending(&self) -> Result<Vec<DedupWarning>, DomainError>;
    async fn dismiss(&self, id: Uuid) -> Result<(), DomainError>;
    async fn mark_merged(&self, id: Uuid) -> Result<(), DomainError>;
    async fn exists_pair(&self, a: Uuid, b: Uuid) -> Result<bool, DomainError>;
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn dedup_warning_status_equality() {
        assert_eq!(DedupWarningStatus::Pending, DedupWarningStatus::Pending);
        assert_ne!(DedupWarningStatus::Pending, DedupWarningStatus::Dismissed);
    }

    #[test]
    fn new_dedup_warning_holds_fields() {
        let a = Uuid::new_v4();
        let b = Uuid::new_v4();
        let warning = NewDedupWarning {
            resource_id_a: a,
            resource_id_b: b,
            similarity_score: 0.92,
        };
        assert_eq!(warning.resource_id_a, a);
        assert_eq!(warning.resource_id_b, b);
        assert!((warning.similarity_score - 0.92).abs() < f64::EPSILON);
    }
}
